### ISSUE TITLE
FactorsInt in library

### ISSUE THREAD

**Issue by hulpke** [gap-system/gap#2087]
This is prompted by a forum question which claimed that `AbelianInvariants`can be held back by using `FactorsInt` instead of `Factors`.

[BTW: I have not found where `AbelianInvariants` calls `FactorsInt`, only calls to `Factors`, but that is not a disproof.]

There are about 2 dozen places in the library in which `FactorsInt` is called. This being a global function, it cannot be overridden by a package.

While there might be borderline cases that really need the Rho factorization, it seems to me that the main difference is a minor speedup in that `FactorsInt` does not need to construct the default ring and redispatch (as `Factors(n)` would), but that could be resolved by another method installation for condition `IsInt`.

Is this really time critical? Is there any reason to not replace the `FactorsInt` calls in the library (short of the factorization routine) with `Factors`? (I can do so)

**Comment by fingolfin** [comment]
The library method in lib/grp.gi:631 does so:
```gap
InstallMethod( AbelianInvariants,
    "generic method for groups",
    [ IsGroup ],
    function ( G )
...
    # the parent of this will be G
    cmm := DerivedSubgroup(G);
    for p  in Set( FactorsInt( Size( G ) ) )  do
...
            if r <> 1  then
                Add( ranks, Length(FactorsInt(r)) );
            fi;
...
```
Many other functions also do, e.g. this one for `Exponent`:
```gap
InstallMethod( Exponent,
    "generic method for finite groups",
    [ IsGroup and IsFinite ],
function(G)
  local exp, primes, p;
  exp := 1;
  primes := Set(FactorsInt(Size(G)));
  for p in primes do
    exp := exp * Exponent(SylowSubgroup(G, p));
  od;
  return exp;
end);
```

**Comment by fingolfin** [comment]
Also note that there used to be (and perhaps still are) places  in our that called `Factors` or `FactorsInt` and shouldn't have called either -- e.g. to test whether a number is a primer power, it is *not* necessary to compute a full factorization, and `IsPrimePowerInt` resp. `SmallestRootInt` are *much* faster for this (and get even faster with PR #2061 resp. #2086).

**Comment by hulpke** [comment]
@fingolfin 
Can I interpret your remarks as in favor of such a change? Unless someone objects I'll do it in a few days.

**Comment by fingolfin** [comment]
Yes, sorry, I am in favor

**Comment by frankluebeck** [comment]
Changing calls like  `FactorsInt(x)`  to `Factors(Integers, x)` sounds sensible, because this way the method selection can decide the best available method. (Currently, `FactorsInt` contains an explicit hook to call `Factors` if the `FactInt` package is loaded, but maybe other packages will provide better methods in the future.) 
A call of `Factors(Integers, x)` instead of just `Factors(x)` can give a significant speedup if this is called frequently with small `x`.

**Comment by fingolfin** [comment]
@stevelinton ironically, the manual contains `PrimesDividingSize` as an example for declaring a custom attribute. As to whether it would be useful: not sure, I am neutral for now. It would generalize the existing `PrimePGroup` attribute, I guess.

**Comment by Stefan-Kohl** [comment]
As far as it concerns overheads when factoring small integers -- the following is what I observe (with FactInt loaded):

    gap> for i in [1..1000] do for j in [1..1000] do FactorsInt(j); od; od; time;  
    4656
    gap> for i in [1..1000] do for j in [1..1000] do Factors(j); od; od; time;
    1348
    gap> for i in [1..1000] do for j in [1..1000] do Factors(Integers,j); od; od; $
    1008

**Comment by fingolfin** [comment]
This issues was actually resolved by PR #2057, we almost always use `Factors(Integers,j)` now
